File: src/cache.rs
use std::collections::{HashSet, BTreeMap};
use std::sync::{Arc, Mutex};
use std::time::SystemTime;

use crate::ldap_filter;
use crate::ldap_wrapper::{LdapW, get_all_values};
use crate::my_err::{Result, MyErr};
use crate::my_types::{Mright};
use crate::remote_query::parse_sql_url;


pub type SharedExpirableCache<C> = Arc<Mutex<Option<(SystemTime, Arc<C>)>>>;

type RemoteToSgroupIds = BTreeMap<String, HashSet<String>>;

#[derive(Clone, Default)]
pub struct AllCaches {
    pub remote_to_sgroup_ids: SharedExpirableCache<RemoteToSgroupIds>,
}

pub fn clear<C>(cache: &SharedExpirableCache<C>) {
    *cache.lock().unwrap() = None;
}

fn get<C>(cache: &SharedExpirableCache<C>) -> Option<(SystemTime, Arc<C>)> {
    (*cache.lock().unwrap()).clone()
}

pub fn set<C>(cache: &SharedExpirableCache<C>, data: Arc<C>) {
    *cache.lock().unwrap() = Some((SystemTime::now(), data));
}

pub async fn get_remote_to_sgroup_ids(ldp: &mut LdapW<'_>, cache: &SharedExpirableCache<RemoteToSgroupIds>) -> Result<Arc<RemoteToSgroupIds>> {
    Ok(if let Some((_, r)) = get(&cache) {
        r
    } else {
        let attr = Mright::Member.to_attr_synchronized();
        let mut map = btreemap![];
        for entry in ldp.search_sgroups(&ldap_filter::presence(&attr), vec![&attr], None).await? {
        let sgroup_id = ldp.config.dn_to_sgroup_id(&entry.dn).ok_or(MyErr::Msg("internal error".to_owned()))?;
            let url = get_all_values(entry.attrs).pop().ok_or(MyErr::Msg("internal error".to_owned()))?;
            let remote = parse_sql_url(&url)?.ok_or(MyErr::Msg("internal error".to_owned()))?;
        map.entry(remote.remote_cfg_name).or_insert(hashset![]).insert(sgroup_id);
        }
        let r = Arc::new(map);
        set(&cache, Arc::clone(&r));
        r
    })
}

pub fn clear_all(all_caches: &AllCaches) {
    clear(&all_caches.remote_to_sgroup_ids);
}

File: server/cache.ts
import * as ldpSgroup from './ldap_sgroup_read_search_modify'
import ldap_filter from "./ldap_filter";
import { internal_error } from "./helpers";
import { dn_to_sgroup_id } from "./dn";
import { mono_attrs } from "./ldap_helpers";
import { hMright, MyMap, Option } from "./my_types";
import { parse_remote_query } from './api_get';
import conf from './conf';
import { Periodicity } from './periodicity';

type PeriodicityToSgroupIds = MyMap<Periodicity, Set<string>>;

const _all_caches = {
    periodicity_to_sgroup_ids: undefined as Option<PeriodicityToSgroupIds>,
}

export const get_periodicity_to_sgroup_ids = async () => (
    (_all_caches.periodicity_to_sgroup_ids ??= await get_periodicity_to_sgroup_ids_())
)

async function get_periodicity_to_sgroup_ids_() {
    const attr = hMright.attr_synchronized;
    const map: PeriodicityToSgroupIds = {};
    const remote_group_filter = ldap_filter.and([ conf.ldap.group_filter, ldap_filter.presence(attr) ])
    for (const entry of await ldpSgroup.search_sgroups(remote_group_filter, [attr], undefined)) {
        const sgroup_id = dn_to_sgroup_id(entry.dn) ?? internal_error();
        const url = mono_attrs(entry)[attr] ?? internal_error()
        const remote = parse_remote_query(url) ?? internal_error();
        const periodicity: Periodicity = remote.forced_periodicity ?? conf.remotes[remote.remote_cfg_name]?.periodicity ?? internal_error()
        ;(map[periodicity] ??= new Set()).add(sgroup_id);
    }
    return map;
}

export function clear_all() {
    _all_caches.periodicity_to_sgroup_ids = undefined
}

